frame/contracts/src/migration.rs

```diff
@@ -24,59 +27,117 @@ use frame_support::{
 	pallet_prelude::*,
 	traits::{ConstU32, Get, OnRuntimeUpgrade},
 };
-use sp_std::{marker::PhantomData, prelude::*};
+use sp_std::marker::PhantomData;
+
+#[cfg(feature = "try-runtime")]
+use sp_std::prelude::*;
 
 const PROOF_ENCODE: &str = "Tuple::max_encoded_len() < Cursor::max_encoded_len()` is verified in `Self::integrity_test()`; qed";
 const PROOF_DECODE: &str =
 	"We encode to the same type in this trait only. No other code touches this item; qed";
-const PROOF_EXISTS: &str = "Required migration not supported by this runtime. This is a bug.";
+
+fn invalid_version(version: StorageVersion) -> ! {
+	panic!("Required migration {version:?} not supported by this runtime. This is a bug.");
+}
 
 pub type Cursor = BoundedVec<u8, ConstU32<1024>>;
-type Migrations<T> = (v9::Migration<T>,);
+type Migrations<T> = (NoopMigration<7>, NoopMigration<8>, v9::Migration<T>);
 
-enum IsFinished {
+/// IsFinished describes whether a migration is finished or not.
+pub enum IsFinished {
 	Yes,
 	No,
 }
 
-trait Migrate<T: Config>: Codec + MaxEncodedLen + Default {
+/// A trait that allows to migrate storage from one version to another.
+/// The migration is done in steps. The migration is finished when
+/// `step()` returns `IsFinished::Yes`.
+pub trait Migrate: Codec + MaxEncodedLen + Default {
+	/// Returns the version of the migration.
 	const VERSION: u16;
 
+	/// Returns the maximum weight that can be consumed in a single step.
 	fn max_step_weight() -> Weight;
 
-	fn step(&mut self) -> (IsFinished, Option<Weight>);
+	/// Process one step of the migration.
+	/// Returns whether the migration is finished and the weight consumed.
+	fn step(&mut self) -> (IsFinished, Weight);
```

Comment:
It was not obvious to me what the none value was for the Option<Weight>, we can put it back if you feel it's a better API. since you can easily default to max_step_weight() within a given Migrate step, I think it's better to make it non optional

Reply:
I think it's fine for now, tuples of 2 are actually pretty nice imo, they are less verbose to destructure, and it's one less type definition to search for and read!